packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -177,25 +265,130 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
         getLocaleDictionary(),
       ]);
 
-      const versionsWithMeta = results.map((version) => {
-        return {
-          ...version,
-          meta: {
-            unknownAttributes: getSchemaAttributesDiff(
-              version.schema,
-              strapi.getModel(params.contentType).attributes
-            ),
+      type EntryToPopulate =
+        | {
+            documentId: string;
+            locale: string | null;
+          }
+        | { id: Data.ID }
+        | null;
+
+      /**
+       * Get an object with two keys:
+       * - results: an array with the current values of the relations
+       * - meta: an object with the count of missing relations
+       */
+      const buildRelationReponse = async (
+        values: EntryToPopulate[],
+        attributeSchema: Struct.SchemaAttributes[keyof Struct.SchemaAttributes]
+      ): Promise<{ results: any[]; meta: { missingCount: number } }> => {
+        return (
+          values
+            // Until we implement proper pagination, limit relations to an arbitrary amount
+            .slice(0, 25)
+            .reduce(
+              async (currentRelationDataPromise, entry) => {
+                const currentRelationData = await currentRelationDataPromise;
+
+                // Entry can be null if it's a toOne relation
+                if (!entry) {
+                  return currentRelationData;
+                }
+
+                const isNormalRelation =
+                  attributeSchema.type === 'relation' &&
+                  attributeSchema.relation !== 'morphToOne' &&
+                  attributeSchema.relation !== 'morphToMany';
+
+                /**
+                 * Adapt the query depending on if the attribute is a media
+                 * or a normal relation. The extra checks are only for type narrowing
+                 */
+                let relatedEntry;
+                if (isNormalRelation) {
+                  if ('documentId' in entry) {
+                    relatedEntry = await strapi
+                      .documents(attributeSchema.target)
+                      .findOne(entry.documentId, { locale: entry.locale || undefined });
+                  }
+                } else if ('id' in entry) {
```

Comment:
what's the difference here? 🤔

Reply:
documentId is needed for relations, id is needed for media assets. I added a comment